### PR TITLE
Folding of unary `-`

### PR DESCRIPTION
**Pull request by romainfrancois**

### CODE REVIEW

File: inst/include/dplyr/hybrid/Expression.h [hunk 1]
```diff
@@ -173,24 +173,49 @@ class Expression {
   // is the i-th argument a scalar int
   inline bool is_scalar_int(int i, int& out) const {
     SEXP val = values[i];
-    if (Rf_length(val) != 1) return false;
+    bool unary_minus = false;
+
+    // unary minus
+    if (TYPEOF(val) == LANGSXP && Rf_length(val) == 2 && CAR(val) == Rf_install("-")) {
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: If we have to go there, I'd prefer actually not handle `-` at all, seems expensive to have to evaluate `-` in the quosure env, then test it against `base::-`

File: inst/include/dplyr/hybrid/Expression.h [hunk 2]
```diff
@@ -173,24 +173,49 @@ class Expression {
...
     switch (TYPEOF(val)) {
     case INTSXP:
     {
+      if (Rf_length(val) != 1) return false;
       int value = INTEGER(val)[0];
       if (IntegerVector::is_na(value)) {
         return false;
       }
-      out = value;
+      out = unary_minus ? -value : value;
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: I suppose you mean `-INT_MAX` right ? then it's fine: 

```r
> Rcpp::evalCpp("-INT_MAX")
[1] -2147483647
```

File: inst/include/dplyr/hybrid/Expression.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: If we have to go there, I'd prefer actually not handle - at all, seems expensive to have to evaluate - in the quosure env, then test it against base::-

File: inst/include/dplyr/hybrid/Expression.h [hunk 3]
```diff
@@ -173,24 +173,49 @@ class Expression {
...
     }
     case REALSXP:
     {
+      if (Rf_length(val) != 1) return false;
       double value = REAL(val)[0];
       if (NumericVector::is_na(value)) {
         return false;
       }
-      out = Rcpp::internal::r_coerce<REALSXP, INTSXP>(value);
+      value = Rcpp::internal::r_coerce<REALSXP, INTSXP>(value);
```
[the remark this answers is not in the document]
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: It gives `NA` so we can just swap the two lines:

```r
> Rcpp::evalCpp( 'Rcpp::internal::r_coerce<REALSXP, INTSXP>(::pow(2, 40))')
[1] NA
> Rcpp::evalCpp( 'Rcpp::internal::r_coerce<REALSXP, INTSXP>(NA_REAL)')
[1] NA
```

File: inst/include/dplyr/hybrid/Expression.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (krlmlr) [reply inst/include/dplyr/hybrid/Expression.h]: Yeah, overriding `-` looks like a particularly efficient way of shooting yourself in the foot. Users will have more severe problems that that afterwards.

File: inst/include/dplyr/hybrid/Expression.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (krlmlr) [reply inst/include/dplyr/hybrid/Expression.h]: I meant `INT_MIN` but forgot that this is the representation of `NA` .

[the remark this answers is not in the document]
Comment (romainfrancois) [reply inst/include/dplyr/hybrid/Expression.h]: Right I see what you meant, so we're fine.

File: inst/include/dplyr/hybrid/Expression.h [hunk 4]
```diff
@@ -173,24 +173,48 @@ class Expression {
...
+        return false;
+      }
+    }
+
     switch (TYPEOF(val)) {
     case INTSXP:
     {
+      if (Rf_length(val) != 1) return false;
       int value = INTEGER(val)[0];
       if (IntegerVector::is_na(value)) {
```
Comment (krlmlr) [inline inst/include/dplyr/hybrid/Expression.h]: Looks like these few lines are common between `INTSXP` and `REALSXP` now.

**Pull request merged**